Added first draft for event handling guidelines for Android. (#4351)
Azure/azure-sdk: docs/android/design.md

#### Event handling

Android applications commonly need to react to events from the UI or service. The following guidelines apply to client libraries that provide APIs to handle events.

##### Event Handlers

{% include requirement/MUST id="android-event-type %} group related types of events together in an enumeration (see [Enumerations](#Enumerations)) based on Azure Core's `ExpandableStringEnum`. Said enumeration should end with the `EventType` suffix.

{% include requirement/MAY id="android-event-details %} provide additional details about an event to a handler via an instance of a class extending from `Event` in Azure Core. The name for the aforementioned class should end with the `Event` suffix. For example:

```java
class MessagingEvent extends Event<MessagingEventType> {
    private final String threadId;

    MessagingEvent(MessagingEventType eventType, String threadId) {
        super(eventType);
        this.threadId = threadId;
    }

    public String getThreadId() {
        return threadId;
    }

> }

Reviewer: Are these examples of extra properties or is boilerplate that every specific event needs?  If everything needs it, can this live in the base class?

Author: We will remove the base `Event` class but encourage devs to bundle event details together in a class such as this. They will still be able to pass things like `String` directly to a handler

```

{% include requirement/MAY id="android-event-handler-mutable" %} mutate individual event handlers after client instantiation, unlike most client configuration which is required to be immutable.

{% include requirement/MUST id="android-event-handler-registration %} register event handlers on clients via methods whose names start with the `add` prefix and end with the `Handler` suffix. Additionally, said methods must return a unique identifier for the provided handler to facilitate its removal.

{% include requirement/MUST id="android-event-handler-registration-minimum-arguments %} have said methods take at least one argument that is an implementation of the `EventHandler` interface from Azure Core:

```java
public String addOnMessageReceivedHandler(EventHandler<MessageReceivedEvent> handler) {
    ...
    return handlerId;
}
```

{% include requirement/MAY id="android-event-handler-registration-with-event-type %} choose to associate a specific event type to a given event handler during registration instead:

```java
public String addMessagingEventHandler(MessagingEventType eventType, EventHandler<MessagingEvent> handler){
    ...
    return handlerId;
}

> ```

Reviewer: Per our discussion, I'd recommend the second option by the MUST and don't allow the first.

Author: After talking to the architects, we will be going with the first approach only as they feel it's more idiomatic.

{% include requirement/MUST id="android-event-handler-removal %} provide at least one method for unregistering event handlers that takes a unique identifier associated with the handler in question.

{% include requirement/MUST id="android-handling-events %} handle events in methods named with the `on` prefix.

```java
public void onMessageReceived(MessageReceivedEvent event) {
    ...
}
```

```java

> public void onMessagingEvent(MessagingEventType eventType, MessagingEvent event) {

Author: Should we instead use this nomenclature? @JonathanGiles @srnagar 
```suggestion
public void handleMessagingEvent(MessagingEventType eventType, MessagingEvent event) {
```

Author: @JonathanGiles Here's an old question I had that could play a role on how we name the methods we talk about [here](https://github.com/Azure/azure-sdk/pull/4351#discussion_r901193360).

{% include requirement/MUSTNOT id="android-pagination-large-get-iterator" %} expose an iterator over each individual item if getting each item requires a corresponding GET request to the service. One GET per item is often too expensive and so not an action we want to take on behalf of users.

#### Event handling

> Android applications commonly need to react to events from the UI or service. The following guidelines apply to SDKs that expose events to the customer.

Reviewer 2: SDKs -> client libraries

> Android applications commonly need to react to events from the UI or service. The following guidelines apply to SDKs that expose events to the customer.

Reviewer 2: "that expose events to the customer." -> "that provide APIs to handle events".

> ##### Handlers

Reviewer 2: ```suggestion
##### Event Handlers
```

> {% include requirement/MUST id="android-event-unicast-handler" %} declare unicast event handlers as functional interfaces that use the `@FunctionalInterface` annotation.

Reviewer 2: `@FunctionalInterface` is available for Java 8+ versions. Android libraries targeting lower versions of Java will not be able to use this annotation.

> {% include requirement/MUST id="android-event-unicast-handler" %} declare unicast event handlers as functional interfaces that use the `@FunctionalInterface` annotation.

Reviewer 2: We should also provide guidance on when to use unicast vs multicast handlers. While we lean towards multicast handlers to be the default, what are the scenarios when unicast is desirable?

Author: Agreed, I think we could do something like what we do for optional arguments, where if there will be just a finite number of handlers and events you can follow these semantics but if will be more than X or it can grow we tell developers to go with multicast from the get go. What do you think?

Reviewer 3: I think that will lead to confused APIs where end users will not have certainty where to look for the event handling functionality. I would propose that by default always have multicast and have unicast be on a case-by-case approval basis.

> {% include requirement/MAY id="android-event-handler-mutable" %} allow users to add and remove individual event handlers after client instantiation through the `add` and `remove` methods detailed below. This is not considered to be mutating the state of the client.

Reviewer 3: We should consider the builder- vs client-level configuration of event handlers. It is easy to understand why client-level API to add / remove / set event handlers is good, but there exists prior art in our own libraries where we set the event handlers on the builder. For example, the Event Hubs processor APIs have process* methods on the builder, because this forms a fundamental aspect of the processor, and it could be argued that it would be counter-intuitive to use a builder to create a processor, only to configure its fundamental functionality in the subsequently constructed processor client.

The wider question is - do we allow this, and what are the conditions for allowing it? Then, if it is allowed, what is the naming pattern that we use for builder-based event handlers versus client-based event handlers?

{% include requirement/MUST id="android-event-unicast-handler-naming" %} name event handler interfaces using the suffix `Handler`.
> NOTE: The name `EventHandler` is reserved for use in the `azure-core` library.

{% include requirement/MUST id="android-event-unicast-method-naming" %} name event handler methods in unicast handlers using the `on` prefix.

```java
@FunctionalInterface
interface MessageHandler<E extends MessagingEvent> {
    void onMessageReceived(E event);
}
```

{% include requirement/MUST id="android-event-without-properties %} group related types of events together in an enumeration (see [Enumerations](#Enumerations)) based on `azure-core`'s `ExpandableStringEnum`. Said enumeration should end with the `EventType` suffix.

> If you need to provide additional details about an event to a handler, {% include requirement/MUST id="android-event-with-properties %} do so by declaring a class that extends from `Event` in `azure-core` and passing an instance of said class to the handler in question. The name for the aforementioned class should end with the `Event` suffix. For example:

Reviewer 2: `MUST` should be at the beginning

}
```

{% include requirement/SHOULD id="android-group-event-details" %} bundle together event details to pass to an event handler via an instance of a class whose name ends with the `Event` suffix. For example:

```java
public static class MessageReceivedEvent {
    private String messageId;
    private String userId;

    public String getMessageId() {
        return messageId;
    }

    public MessageReceivedEvent setMessageId(String messageId) {
        this.messageId = messageId;

        return this;
    }

    public String getUserId() {
        return userId;
    }

    public MessageReceivedEvent setUserId(String userId) {
        this.userId = userId;

        return this;
    }

> }

Reviewer 3: Because all events are generated by us (and never by users), we should expect that an `*Event` is `@Immutable`. So, no setters should be shown, and rather we should have a ctor taking all required parameters.

> {% include requirement/MUST id="android-event-multicast" %} expose multicast event handlers using the `MulticastEventCollection` type from `azure-core`. This type contains an `addEventHandler` method which accepts the event alongside an implementation of the `EventHandler` interface to handle it and returns a UUID string identifier which can be used to remove the handler via the `removeEventHandler` method. Here's an example:

Reviewer 2: `MulticastEventCollection` should not be directly visible to the customer. It must only be used by library developers. So, we have to rephrase this "expose multicast event handlers using `MulticastEventCollection`".

```

{% include requirement/MAY id="android-event-details-other-classes" %} instead provide other types of objects to event handlers when grouping event details together is not necessary. For example:

```java
public void addOnAlertTriggeredHandler(EventHandler<String> handler) {
    ...
}

...

EventHandler<String> alertTriggeredHandler = new EventHandler<String>() {
    @Override

>     public void handle(String timestamp) {

Reviewer 3: `timestamp` as the parameter name is probably not the best choice - otherwise you should probably change the EventHandler to be Long rather than String. If you want to keep it as String, lets find a better parameter name, e.g. 'filename', 'message', 'userInput', 'alertMessage', etc.

        ...
    } 
}

...

client.addOnAlertTriggeredHandler(alertTriggeredHandler);

> ```

Reviewer 3: It took me a few reads of the text to fully understand what was intended, even though I remember discussing it a long time ago. "MAY instead provide other types of objects to event handlers when grouping event details together is not necessary" might be better written as "MAY use existing types, rather than a new Event type, when an existing type provides all necessary data necessary to the user. Note that this prevents future growth to provide additional information to the user, so do not do this without much consideration."

> }

Reviewer 2: Creating a new type is not necessary.

>     public void handleMessagingEvent(MessagingEvent event) {

Reviewer 2: Doesn't this also require the eventType?

Author: I wrote two overloads for this in `MulticastEventCollection`: `handleEvent(T eventType, E event)` and `handleEvent(E event)`. The first one is the original method and the second one calls it after extracting the event type from `event`, but only if the object is a subclass of `Event`, else it throws. We could remove the second overload and simply work with the first one I mentioned if you think it's better for now. I thought that since `eventType` is already a member of `Event`, some users might think it's unnecessary to extract it manually to call `handleEvent()`.

Also, if you're thinking: "Is there any occasion where we won't pass an instance `Event` to the handler?", I thought not all events might carry details other than the type, so would it not feel unnecessary to box the type in `Event` for these cases? In such a case, I guess we could have a method to handle an event that only receives its type? I don't know, I need to ponder a bit more on this.

{% include requirement/MUSTNOT id="android-pagination-large-get-iterator" %} expose an iterator over each individual item if getting each item requires a corresponding GET request to the service. One GET per item is often too expensive and so not an action we want to take on behalf of users.

#### Event handling

Android applications commonly need to react to events from the UI or service. The following guidelines apply to client libraries that provide APIs to handle events.

##### Event Handlers

> {% include requirement/MAY id="android-event-handler-mutable" %} mutate individual event handlers after client instantiation, unlike most client configuration which is required to be immutable.

Reviewer 3: ```suggestion
{% include requirement/MAY id="android-event-handler-mutable" %} allow users to add and remove individual event handlers after client instantiation through the `add` and `remove` methods detailed below. This is not considered to be mutating the state of the client.
```

> {% include requirement/MUST id="android-event-handler-collection" %} use Azure Core's `EventHandlerCollection` internally to keep track of registered event handlers and their associations with different event types within a client.

Reviewer 3: This isn't strictly necessary. It is recommended though.. I would change it to ensure that the given code sample demonstrates how this is to be used.

> {% include requirement/MUST id="android-event-types" %} declare event types to associate event handlers with as package private client constants of the type `String`.

Reviewer 3: We should move this guidance (and the previous one) from the API design documentation and into the implementation documentation.

> {% include requirement/MUST id="android-event-handler-registration" %} register event handlers on clients via methods whose names start with the `addOn` prefix and end with the `Handler` suffix.

Reviewer 3: I'm not convinced the `On` adds much value, e.g. is `addOnKeyTypedHandler` better than `addKeyTypedHandler`? The `On` makes more sense if we didn't enforce the `Handler` suffix - `addOnKeyTyped` requires the `On`, because we've dropped `Handler`. What is more useful for developers? We have:

* `addOnKeyTypedHandler`
* `addKeyTypedHandler`
* `addOnKeyTyped`

Author: When I originally wrote this I thought it would be good to be overtly explicit with the names to avoid some awkward names. For example, if the event method in a client would be called `onClick()`, I would rather have `addOnClickHandler()` than `addClickHandler()` or `addOnClick()`. I get that for methods that use the past form of a verb that might not be necessary, with things like `onClick()` or `onCreate()` being a standard for Android this felt like the right thing to do. We could drop the `Handler` suffix for something like `addOnKeyTyped()` but I am not sure if it's better to just have a single guidance than two different ones based on some condition.

Reviewer 3: I don't strongly hold an opinion on this - the main thing is the consistency that we want to derive from it. `addOn*Handler` is fine with me, but I'm open to discussion on shortening this in various ways.

> {% include requirement/MUST id="android-event-handler-registration-arguments" %} have said methods take one argument that is an implementation of the `EventHandler` interface from Azure Core. This argument must be named `handler`.

Reviewer 3: ```suggestion
{% include requirement/MUST id="android-event-handler-registration-arguments" %} define event handler methods as taking a single argument of type `EventHandler` (from Azure Core). This argument must be named `handler`.
```
